### DISCUSSION TITLE
How to properly combine NAT traversal with peer discovery?

### DISCUSSION THREAD

**Discussion by Wiezzel** [libp2p/rust-libp2p#3673] [Q&A] [answered]
### General problem statement

I would like to implement a p2p network, using libp2p, with the following features:
1. There's at least one bootnode, with publicly known address and peer ID. It also acts as a relay server.
2. Other nodes could be listening on public address, but don't necessary have to. 
3. Given only the peer ID of some peer in the network, any node should be able to establish a connection to that peer.
  a. A direct connection is preferred, if either of the two nodes has a public address,
  b. If none of the nodes has a public address, they should attempt a DCUtR,
  c. If DCUtR fails, fall back to a relayed connection.
  
As far as I understand, I need to include the following behaviours in my swarm:
1. `relay::v2::client` – for relaying,
2. `dcutr` – for NAT hole punching,
3. `identify` & `kad` – for finding peers.

### Questions

1. Do I miss any behaviour in the list above?
2. What is the proper sequence of actions to execute on node startup? I mean like 'call method x of behaviour Y', 'listen for event Z', etc. I would like to make sure here that a newly started node is properly registered with the relay server and it's relayed address is discoverable by others.
3. What is the proper sequence of actions to find a peer's address given its peer ID? Is it enough to call `.get_closest_peers(peer_id)` and wait for `OutboundQueryProgressed`? But how do I extract the address?

**Comment by dariusc93** [comment, marked as answer]
>     1. Do I miss any behaviour in the list above?
 I would also suggest `Ping` behaviour too otherwise you may timeout get timeout or disconnection sometime after the connection. May also suggest `AutoNat` to determine the peer nat status too.
 
>     2. What is the proper sequence of actions to execute on node startup? I mean like 'call method x of behaviour Y', 'listen for event Z', etc. I would like to make sure here that a newly started node is properly registered with the relay server and it's relayed address is discoverable by others.
The way I do it is i would make sure the relay server is connected to and tracking the `ConnectionId` (you dont really have to do this though), and possibly wait for `Identify` to come in. May want to check the protocol to be sure that they have the relay server protocol. Once that is done, you would want to proceed to listen on the relay circuit (eg `/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN/p2p-circuit`) and make sure it listened successfully (you can either do this while polling swarm or in your own behaviour - depending on your use case) and you should be able to proceed from there, although you may want to check events from the relay client behaviour for any rejection of the reservation, etc.

>     3. What is the proper sequence of actions to find a peer's address given its peer ID? Is it enough to call `.get_closest_peers(peer_id)` and wait for `OutboundQueryProgressed`? But how do I extract the address?

When you call `Kademlia::get_closest_peers`, you would wait for `QueryResult::GetClosestPeers`, then you could get the addresses from kbucket (originally you could from  `Kademlia::addresses_of_peer`, but that is deprecated). 

The way I would do it is i would check track the connections and return a list of addresses associated to that peer. 

Eg, call `Kademlia::get_closest_peers`, track the `QueryId` or the `Key` while polling the swarm, waiting for `QueryResult::GetClosestPeers(Ok(GetClosestPeersOk { .. }))` and once received (assuming the query is finished), check the list for the peer in question and send the results via channel (in my case is usually a oneshot channel but if youre using or tracking the `Key` instead, I would suggest either vector of channels or a broadcast channel assigned to the `Key` in case you make multiple calls for the same key and want the results same results for all those calls) then would make calls to fetch the info related to the peer thats found from my behaviour that handles storing that info, and if that info isnt there it would fail, or if it is, it would return successfully.
